Automattic/wordpress-activitypub
# Add replies collection

Add [replies collection](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-replies) to posts and federated comments.

**Solves Issue** https://github.com/Automattic/wordpress-activitypub/issues/138

### Description
The replies collection offer a neat way to sync replies over multiple instances. Mostly on demand and sometimes also client side on demand. They are not yet frequently used because of the moderation issues that arise with them. But since WordPress already got a good approve-system for comments this should not be too problematic here.

### Rest Endpoint
To make sure the ID's of the collection and the collections page are resolveable.
It also adds a new rest endpoint `/activitypub/1.0/(posts|comments)/<id>/replies` which returns a `Collection`. Adding the parameter `?page=0` returns the CollectionPage directly.
I decided not to use something like `<the_permalink>/replies` because it is only needed for `activity+json` and has no frontend view. Seemed consistent to me.

### Unit tests:
I started writing new unit tests, but was not sure how deep to test this. I'm happy for some discussions.

### Manual testing:
Write federated comments for a post, or receive comments and make sure those comments are approved.
When querying the ActivbityPub's JSON representation of the post the replies key exists, e.g.:

```json
...
 "replies": {
    "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
    "type": "Collection",
    "first": {
      "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=0",
      "type": "CollectionPage",
      "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
      "items": [
        "https://mz.lan/comments/8be36040-5a02-4335-917f-4c5dcc988e1f",
        "https://mz.lan/comments/de26c344-8d7d-400f-a8f5-5b89307bb72a",
        "https://mz.lan/comments/abce32af-a59e-41c6-8b47-a8925980f1ef",
        "https://wp.lan/?c=13",
        "https://mz.lan/comments/fa3d2ee2-147b-4eee-b682-98b0b7115603"
      ]
    }
  },
...
```

Changing the global option `comments_per_page` to for example "3" would change the response to this:

```json
...
"replies": {
    "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
    "type": "Collection",
    "first": {
      "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=0",
      "type": "CollectionPage",
      "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
      "items": [
        "https://mz.lan/comments/8be36040-5a02-4335-917f-4c5dcc988e1f",
        "https://mz.lan/comments/de26c344-8d7d-400f-a8f5-5b89307bb72a"
      ],
      "next": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1"
    }
  },
...
```

Note that the items only contains 2 IDs. This is because the code currently uses core functions to query the paginated comments and later filters out the local-only ones. Because we have more pages also the key `next` is present.

It (`https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1`) resolves to:

```json
**https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1**
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    {
      "Hashtag": "as:Hashtag"
    }
  ],
  "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1",
  "type": "CollectionPage",
  "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
  "items": [
    "https://mz.lan/comments/abce32af-a59e-41c6-8b47-a8925980f1ef",
    "https://wp.lan/?c=13"
  ],
  "next": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=2"
}
```


**Unapproved or local only comments should never be listed.**

### Discussion points
- This draft has always functions handle both replies for comments and posts rather than splitting up the code in the beginning. I just chose this construct to get to a working prototype and I already fear that there might be good arguments to do it differently.
- The code which gets the federated comment IDs: https://github.com/Menrath/wordpress-activitypub/blob/replies_collection/includes/collection/class-replies.php#L108 These lines are duplicated or have very similar occurences, more than once in the code.
- Maybe use a direct meta-query to get all approved-federated comments in a paginated way.
- I am happy to hear any ideas and thoughts.
- Mastodon for example lists all self-replies on the first page of the collection, and external replies later, this would also be possible, to ensure threads are displayed. Threads are possible now also in WordPress, but maybe not that important.
- Mastodon includes local replies fully, not only the ID. Within WordPress this in theory could get difficult because of the JSON-LD context (different post types with different ActivityPub object types at worst).
